integration-test/test.sh

```diff
@@ -0,0 +1,350 @@
+		local exit_code=$?
+		set -e
+
+		if [ $exit_code -ne 0 ]; then
+			printf '.'
+			sleep 5
+		else
+			log "Authenticated to docker engine at $addr."
+			# Check if docker.options in public.json took effect
+			local docker_info_out="$(eval $docker_env $docker_cmd 2>&1)"
+			if [[ "$docker_info_out" != *"foo=bar"* ]]; then
+				err "Docker engine label (foo=bar) specified in extension configuration did not take effect."
+				log "docker info output:"
+				log "$docker_info_out"
+				exit 1
+			fi
+			log "Docker configuration took effect."
+			return
+		fi
+	done
+}
+
+wait_for_container() {
+	local host=$1
+	local addr="http://$1:80/"
+
+	log "Waiting for web container on $addr..."
+	local curl_cmd="curl -sILfo/dev/null $addr"
+	echo "+ $curl_cmd"
+	
+	while true; do
+		set +e # ignore errors b/c the following command will retry
+		eval $curl_cmd 2>&1 1>/dev/null
+		local exit_code=$?
+		set -e
+
+		if [ $exit_code -eq 0 ]; then
+			log "Web container is up."
+			return
+		fi
+		printf '.'
+		sleep 5
+	done
+}
+
+validate_extension_version() {
+	local fqdn=$1
+	log "Validating extension version on VM."
+	
+	# Search for file Microsoft.Azure.Extensions.DockerExtension-{version}
+	local prefix="${EXTENSION_PUBLISHER}.${EXTENSION_NAME}-"
+
+	# Find out what version of extension is installed by running
+	# a Docker container with /var/lib/waagent mounted
+	local docker_env="DOCKER_CERT_PATH=\"$(docker_cert_path)\" DOCKER_HOST=\"$(docker_addr $fqdn)\""
+	local docker_cmd="docker --tls run --rm -i -v /var/lib/waagent:/agent busybox ls -1 /agent | grep '^$prefix'"
+
+	echo "+ $docker_env $docker_cmd"
+	local version="$(eval $docker_env $docker_cmd 2>/dev/null | sed "s/^$prefix//g")"
+	is_empty "$version" && {
+		err "Could not locate $EXTENSION_NAME version."
+		exit 1
+	}
+	
+	if [[ "$version" != "$expected_extension_ver" ]]; then
+		err "Wrong $EXTENSION_NAME encountered: '$version' (expected: '$expected_extension_ver')."
+		exit 1
+	fi
+	log "VM has the correct version of $EXTENSION_NAME."
+}
+
+vm_ssh_cmd() {
+	local fqdn=$1
+	echo "ssh -i '$(ssh_key)' ${VM_USER}@${fqdn}"
+}
+
+validate_vm() {
+	local name=$1
+	local fqdn=$(vm_fqdn $name)
+
+	log "Validating $EXTENSION_NAME on VM '$name'"
+	log "    (To debug issues: $(echo $(vm_ssh_cmd $fqdn)))"
+	wait_for_docker $fqdn
+	wait_for_container $fqdn
+	validate_extension_version $fqdn
+
+	log "VM is O.K.: $name."
+	echo
+}
+
+validate_vms() {
+	log "Validating VMs..."		
+	local vms=$(get_vms)
+	for vm in $vms; do
+		validate_vm "$vm"
+	done
+}
+
+main() {
+	check_deps
+	intro
+
+	read -p "Expected $EXTENSION_NAME version in VMs (e.g. 1.0.1512030601): " expected_extension_ver
+	is_empty "$expected_extension_ver" && { err "Empty string passed"; exit 1; }
+
+	check_asm
+	set_subs
+
+	delete_vms
+	create_vms
+	add_extension_to_vms
+	validate_vms
+
+	log "Test run is successful!"
+	log "Cleaning up test artifacts..."
+	delete_vms
+
+	echo
+	log "Success."
+}
+
+
+main
```

Comment:
I totally expect *go* programmers to try to make bash scripts look like something other than a scripting language. It’s an occupational hazard of using well-structured programming languages as your main tool and only occasionally dipping into bash scripting.

bash isn’t a language where everything is a function and there’s a special “start here” function. The more “make it look like a *real* programming language” syntactic sparkles one puts into a shell script, the more one is tempted to forget one is working in a scripting environment. It’s like using macros to make C look like ALGOL-68; S. R. Bourne did it when he wrote the origin /bin/sh, but we’ve been wagging our fingers at him ever since.

I’m fine with building functions to neatly organize a block of script statements, although if that function is invoked precisely once I question the value. But there’s no organizational or abstractional value to defining a function called “main()” and invoking it precisely once as the last line of the script, just after the place where you defined the function in the first place.

The whole point is that it is **not** at the same indent level as the rest of the script. All of that other stuff was defining what you're going to do; when you get to this point, you're actually doing it.